### ISSUE TITLE
Sign in to dropbox fails with error: "Precondition failed: " must match pattern "\A(?:(/(.|[\r\n])*|id:.*)|(rev:[0-9a-f]{9,})|(ns:[0-9]+(/.*)?))\z": file SwiftyDropbox/StoneValidators.swift, line 9"

### ISSUE THREAD

**Issue by gerocar** [dropbox/SwiftyDropbox#328]
I'm setting up a new app in swift which I want to connect to dropbox to save and read files.

I imported SwiftyDropbox via cocoa pods.

When I launch the app in the simulator of Xcode, the sign in screen from dropbox for the new app is shown and I'm able to sign in and approve the access. 

But when the dropbox app wants to return to my app, I get the following error 

`2021-07-31 07:03:21.510744+0200 U*******r[1188:22933] [Presentation] Presenting view controller <UIAlertController: 0x7fac4703c800> from detached view controller <U*******r.HomeVC: 0x7fac47810a00> is discouraged.
2021-07-31 07:08:34.548278+0200 U*******r[1188:22933] -canOpenURL: failed for URL: "dbapi-2://1/connect?k=j*************w&s=" - error: "The operation couldn’t be completed. (OSStatus error -10814.)"
2021-07-31 07:08:34.549180+0200 U*******r[1188:22933] -canOpenURL: failed for URL: "dbapi-8-emm://1/connect?k=j*************w&s=" - error: "The operation couldn’t be completed. (OSStatus error -10814.)"
2021-07-31 07:08:35.148832+0200 UmzugManager[1188:22933] [Presentation] Attempt to present <UIAlertController: 0x7fac4904e200> on <UINavigationController: 0x7fac4781fc00> (from <U*******r.HomeVC: 0x7fac47810a00>) whose view is not in the window hierarchy.
Success! User is logged into DropboxClientsManager.
Precondition failed: " must match pattern "\A(?:(/(.|[\r\n])*|id:.*)|(rev:[0-9a-f]{9,})|(ns:[0-9]+(/.*)?))\z": file SwiftyDropbox/StoneValidators.swift, line 9
2021-07-31 07:10:50.355491+0200 U*******r[1188:22933] 
stall(lldb) `

The setup in AppDelegate is as follows:

`import UIKit
import SwiftyDropbox

@main
class AppDelegate: UIResponder, UIApplicationDelegate {
    let dropboxAppKey = "j***********w"
    var rootViewController: UINavigationController?

    func application(_ application: UIApplication, didFinishLaunchingWithOptions launchOptions: [UIApplication.LaunchOptionsKey: Any]?) -> Bool {
        DropboxClientsManager.setupWithAppKey(dropboxAppKey)
        return true
    }

    // MARK: UISceneSession Lifecycle

    func application(_ application: UIApplication, configurationForConnecting connectingSceneSession: UISceneSession, options: UIScene.ConnectionOptions) -> UISceneConfiguration {
        return UISceneConfiguration(name: "Default Configuration", sessionRole: connectingSceneSession.role)
    }

    func application(_ application: UIApplication, didDiscardSceneSessions sceneSessions: Set<UISceneSession>) {
    }
    
    func application(_ app: UIApplication, open url: URL, options: [UIApplication.OpenURLOptionsKey : Any] = [:]) -> Bool {
     let oauthCompletion: DropboxOAuthCompletion = {
          if let authResult = $0 {
              switch authResult {
              case .success:
                  print("Success! User is logged into DropboxClientsManager.")
              case .cancel:
                  print("Authorization flow was manually canceled by user!")
              case .error(_, let description):
                  print("Error: \(String(describing: description))")
              }
          }
        }
        let canHandleUrl = DropboxClientsManager.handleRedirectURL(url, completion: oauthCompletion)
        return canHandleUrl
    }
}
`

And in SceneDelegate:

`import UIKit
import SwiftyDropbox

class SceneDelegate: UIResponder, UIWindowSceneDelegate {

    var window: UIWindow?
    let app = UIApplication.shared.delegate as! AppDelegate
    var rootVC: UINavigationController?

    func scene(_ scene: UIScene, willConnectTo session: UISceneSession, options connectionOptions: UIScene.ConnectionOptions) {
        guard let _ = (scene as? UIWindowScene) else { return }
        rootVC = self.window?.rootViewController as? UINavigationController
        app.rootViewController = rootVC    }

    func sceneDidDisconnect(_ scene: UIScene) {
    }

    func sceneDidBecomeActive(_ scene: UIScene) {
    }

    func sceneWillResignActive(_ scene: UIScene) {
    }

    func sceneWillEnterForeground(_ scene: UIScene) {
    }

    func sceneDidEnterBackground(_ scene: UIScene) {
    }
    
    func scene(_ scene: UIScene, openURLContexts URLContexts: Set<UIOpenURLContext>) {
         let oauthCompletion: DropboxOAuthCompletion = {
          if let authResult = $0 {
              switch authResult {
              case .success:
                  print("Success! User is logged into DropboxClientsManager.")
              case .cancel:
                  print("Authorization flow was manually canceled by user!")
              case .error(_, let description):
                  print("Error: \(String(describing: description))")
              }
          }
        }
        
        for context in URLContexts {
            if DropboxClientsManager.handleRedirectURL(context.url, completion: oauthCompletion) { break }
        }
    }
    
}
`

For the connection to the dropbox I'm using the following method in the HomeVC-class, referred to in the above quoted error message:

`func connectDropbox(withRequest request: Dropboxrequest) {
        self.dropboxrequest = request
        DropboxClientsManager.authorizeFromController(UIApplication.shared, controller: self, openURL: {(url: URL) -> Void in UIApplication.shared.open(url, options: [:], completionHandler: nil)})
    }
    `
Has anyone an idea what goes wrong?
Thanks for your support.

**Comment by greg-db** [comment]
It looks like there are a few things happening here, so I'll attempt to address each of them individually:

> Presenting view controller `UIAlertController` from detached view controller is discouraged.

This is a message from iOS, more about how your view controllers are being managed, rather than the Dropbox API/SDK itself. I recommend referring to Apple's documentation for information on managing view controllers. There are also some posts about this online, like [this one](https://stackoverflow.com/questions/33366133/getting-warning-like-presenting-view-controllers-on-detached-view-controllers-is), which may offer some guidance. 

> -canOpenURL: failed for URL: "dbapi-2://1/connect?k=j*************w&s=" - error: "The operation couldn’t be completed. (OSStatus error -10814.)"
> -canOpenURL: failed for URL: "dbapi-8-emm://1/connect?k=j*************w&s=" - error: "The operation couldn’t be completed. (OSStatus error -10814.)" 

The canOpenURL 10814 errors themselves are expected if the official Dropbox app isn't installed, so they're safe to ignore. (The SDK checks if the official app is installed, and would use that for the authorization flow if it is. If it isn't though, it will fall back to an in-app flow.)

> Attempt to present `UIAlertController` on `UINavigationController` whose view is not in the window hierarchy

This looks similar to the first issue. In any case, I'm not sure which alert this is actually referring to anyway, since I don't see any alert usage in your code here. You'll need to debug how you're managing your views.

> Success! User is logged into DropboxClientsManager.

This indicates that the Dropbox app authorization flow succeeded anyway.

> Precondition failed: " must match pattern "\A(?:(/(.|[\r\n])*|id:.*)|(rev:[0-9a-f]{9,})|(ns:[0-9]+(/.*)?))\z": file SwiftyDropbox/StoneValidators.swift, line 9

This should indicate that you're making a Dropbox API call with an invalid parameter value. Specifically, you seem to be providing an empty string to a call that does not accept an empty string. I don't see code here to make a Dropbox API call after the authorization flow though. Can you share that piece of code?

**Comment by gerocar** [comment]
Dear greg-db,

thanks for the splendid support.

As you indicated, I could successfully eliminate all messages in reference with the dropbox-app availability, by installing this app on my device.

Also your last hint referring to "Precondition failed: ..." helped. I misinterpreted the whole messages as one in conjunction with the authorization process. In fact, this wasn't so. The last error  message was caused by a file download request with an empty path after the authorization.,
